ENH: Add pipe method

TomAugspurger opened the pull request:
Closes https://github.com/pydata/pandas/issues/10129

In the dev meeting, we settled on the following:
- `.pipe` will **not** include a check for `__pipe_func__` on the function passed in.
- To avoid messiness with lambdas when a function takes the DataFrame other than in the first position, users can pass in a `(callable, data_keyword)`  argument to `.pipe` (thanks @mwaskom)

``` python
   import statsmodels.formula.api as sm

   bb = pd.read_csv('data/baseball.csv', index_col='id')

   (bb.query('h > 0')
      .assign(ln_h = lambda df: np.log(df.h))
      # sm.possion expects `formula, data`
      .pipe((sm.poisson, 'data'), 'hr ~ ln_h + year + g + C(lg)')
      .fit()
      .summary()
   )
## -- End pasted text --
Optimization terminated successfully.
         Current function value: 2.116284
         Iterations 24
Out[1]:
<class 'statsmodels.iolib.summary.Summary'>
"""
                          Poisson Regression Results
==============================================================================
Dep. Variable:                     hr   No. Observations:                   68
Model:                        Poisson   Df Residuals:                       63
Method:                           MLE   Df Model:                            4
Date:                Tue, 02 Jun 2015   Pseudo R-squ.:                  0.6878
Time:                        20:57:27   Log-Likelihood:                -143.91
converged:                       True   LL-Null:                       -460.91
                                        LLR p-value:                6.774e-136
===============================================================================
                  coef    std err          z      P>|z|      [95.0% Conf. Int.]
-------------------------------------------------------------------------------
Intercept   -1267.3636    457.867     -2.768      0.006     -2164.767  -369.960
C(lg)[T.NL]    -0.2057      0.101     -2.044      0.041        -0.403    -0.008
ln_h            0.9280      0.191      4.866      0.000         0.554     1.302
year            0.6301      0.228      2.762      0.006         0.183     1.077
g               0.0099      0.004      2.754      0.006         0.003     0.017
===============================================================================
"""
```

Thanks everyone for the input in the issue.

---

This is mostly ready. What's a good name for the argument to `pipe`? I have `func` right now, @shoyer you had `target` in the issue thread. I've been using `target` as the keyword expecting the data, i.e. where the DataFrame should be pipe to.

The tests are extremely minimal... but so is the implementation. Am I missing any obvious edge-cases?

We'll see how this goes. I don't think I push `.pipe` as a protocol at all in the documentation, though we can change that in the future. We should be forwards-compatible if we do ever go down the `__pipe_func__` route.

TomAugspurger committed:
ENH: this is a pipe

jreback commented on doc/source/basics.rst:
```diff
@@ -624,6 +624,66 @@ We can also pass infinite values to define the bins:
+       )
+
+Pandas encourages the second style. It flows with the rest of pandas
+methods which return DataFrames or Series and are non-mutating by
```
maybe use pure instead?

TomAugspurger replied:
I'm being pedantic, but "pure" implies no-side effects. `df.plot()` is non-mutating, but not pure since it has the side-effect of drawing a plot. I didn't want some functional programming guru to call us out :)

jreback replied:
sure...

shoyer commented on doc/source/basics.rst:
```diff
@@ -624,6 +624,66 @@ We can also pass infinite values to define the bins:
 Function application
 --------------------
 
+There are three main cases for function, depending on what the function
```
There's a missing word after "three main cases for function"

TomAugspurger replied:
-> To apply your own or another library's functions to pandas objects,
you should be aware of the three methods below.

shoyer commented on doc/source/whatsnew/v0.16.2.txt:
```diff
@@ -20,6 +20,47 @@ Check the :ref:`API Changes <whatsnew_0162.api>` before updating.
+their functions. Overall the code is much more readable.
+
+In the example above, the functions ``f``, ``g``, and ``h`` each expected the DataFrame as the first positional argument. When the funciton you wish to apply takes its data anywhere other than the
+frist argument, pass a tuple of ``(funciton, keyword)`` indicating where the DataFrame should flow.For example:
```
frist -> first

need a space after the end of the sentence

shoyer commented on pandas/core/generic.py:
```diff
@@ -2044,6 +2044,55 @@ def sample(self, n=None, frac=None, replace=False, weights=None, random_state=No
+        -----
+
+        Use ``.pipe`` when chaining together functions that operate
+        on or rerturn Series or DataFrames. Instead of writing
```
rerturn -> return

shoyer commented on pandas/core/generic.py:
```diff
@@ -2044,6 +2044,55 @@ def sample(self, n=None, frac=None, replace=False, weights=None, random_state=No
+        func : function
+            function to apply to the %(klass)s
+            ``args``, and ``kwargs`` are passed into ``func``.
+            Alternatively a ``(callable, data_keyword)`` pair where
```
I would say "tuple" not "pair"

shoyer commented on pandas/tests/test_util.py:
```diff
@@ -7,6 +7,7 @@
 import pandas.util
 from pandas.util.decorators import deprecate_kwarg
 import pandas.util.testing as tm
+from pandas import DataFrame
```
looks like this leaked in?

shoyer commented on pandas/core/generic.py:
```diff
@@ -2044,6 +2044,55 @@ def sample(self, n=None, frac=None, replace=False, weights=None, random_state=No
+    def pipe(self, func, *args, **kwargs):
+        if isinstance(func, tuple):
+            func, target = func
+            kwargs[target] = self
```
Should we add some validation logic here to ensure that `target` is not overwriting a key in `kwargs`? Something like this:
```python
if target in kwargs:
    raise ValueError('%s is both the pipe target and a keyword argument' % target)
```

 I'm not entirely sure it's worth complexifying things here.

TomAugspurger replied:
Are you saying maybe the user did something like

```python
df.pipe((sns.violinplot, 'data'), x='x', y='y', data=df)
```

I guess the precedent here is with Python itself raising when you call `f(a=1, a=2)`.

shoyer replied:
yes, that's the case I was thinking about

shoyer commented on doc/source/basics.rst:
```diff
@@ -624,6 +624,66 @@ We can also pass infinite values to define the bins:
+          .pipe(f, arg2=2, arg3=3)
+       )
+
+Pandas encourages the second style. It flows with the rest of pandas
```
I would say something like: "Pandas encourage the second style, which is known as method chaining"

TomAugspurger replied:
-> Pandas encourages the second style, which is known as method chaining.
``pipe`` makes it easy to use your own or another library's functions
in method chains, alongside pandas' methods.

shoyer commented on doc/source/basics.rst:
```diff
@@ -624,6 +624,66 @@ We can also pass infinite values to define the bins:
+
+Pandas encourages the second style. It flows with the rest of pandas
+methods which return DataFrames or Series and are non-mutating by
+default.
```
I would finish this up with: "pipe makes it easy to add functions not built-in to pandas to method chains."

shoyer commented:
A couple of other things that would be nice to highlight in the docs:

1. Let's mention how this was inspired by popular pipe operator `%>%` from R's magrittr package, but the implementation here is explicit and Pythonic. I would encourage reading the source code -- we might even include it in the docs.
2. We should encourage just using `pipe` instead of [monkey patching](http://pandas.pydata.org/pandas-docs/stable/faq.html#ref-monkey-patching). I would consider removing the mention of monkey patching at all -- this is a far better way to go.

jorisvandenbossche commented on doc/source/basics.rst:
```diff
@@ -624,6 +624,66 @@ We can also pass infinite values to define the bins:
+
+   >>> (df.pipe(h),
+          .pipe(g, arg1=1),
+          .pipe(f, arg2=2, arg3=3)
```
I think the comma's at the end of the lines are not correct here?

jorisvandenbossche commented on pandas/core/generic.py:
```diff
@@ -2044,6 +2044,55 @@ def sample(self, n=None, frac=None, replace=False, weights=None, random_state=No
+        Parameters
+        ----------
+        func : function
+            function to apply to the %(klass)s
```
Add a `.` at the end of the line (in html, single line breaks are discarded)

TomAugspurger commented:
> I would consider removing the mention of monkey patching at all

Any objections to removing this section from the docs? I forgot it even existed.

TomAugspurger commented:
> Let's mention how this was inspired by popular pipe operator %>% from R's magrittr package, but the implementation here is explicit and Pythonic. I would encourage reading the source code -- we might even include it in the docs.

The pipe method is inspired by unix pipes and more recently dplyr_ and magrittr_, which
have introduced the popular ``(%>%)`` (read pipe) operator for R_.
The implementation of ``pipe`` here is quite clean and feels right at home in python.
We encourage you to view the source code (``pd.DataFrame.pipe??`` in IPyhton).

---

Ok, addressed all the comments I think (thanks). I removed the section on monkey patching.

The outstanding issue I see is @shoyer's comment about maybe checking whether we're about to clobber a kwarg when the tuple-style is used.

```python
if target in kwargs:
    raise ValueError('%s is both the pipe target and a keyword argument' % target)
```

to catch a case of `df.add(1).pipe((f, 'data'), x=x, y=y, data=df)`. Right now we (silently) replace `kwargs['data']`, The closest parallel I see is writing `f(a=1, a=2)`, which Python catches.

shoyer commented on doc/source/basics.rst:
```diff
@@ -624,6 +624,77 @@ We can also pass infinite values to define the bins:
+The pipe method is inspired by unix pipes and more recently dplyr_ and magrittr_, which
+have introduced the popular ``(%>%)`` (read pipe) operator for R_.
+The implementation of ``pipe`` here is quite clean and feels right at home in python.
+We encourage you to view the source code (``pd.DataFrame.pipe??`` in IPyhton).
```
IPyhton -> IPython

TomAugspurger replied:
Thanks, I should read what I write :/

TomAugspurger committed:
API: catch target kwarg clobbering

TomAugspurger commented:
I put the kwarg clobbering `(df.pipe((f, 'data'), x=1, data=2))` in a second commit: https://github.com/pydata/pandas/commit/968d0aeb217e111692ce904df2c5595ff6140de6

jreback commented:
pls rebase on master and repush had an issue with some builds

TomAugspurger commented:
Travis is green. Are we good with checking whether the `target` is clobbered? And are we OK with raising a `ValueError` instead of `SyntaxError` like `f(a=2, a=3)` does?

jreback commented on doc/source/basics.rst:
```diff
@@ -624,6 +624,77 @@ We can also pass infinite values to define the bins:
+
+1. `Tablewise Function Application`_: :meth:`~DataFrame.pipe`
+2. `Row or Column-wise Function Application`_: :meth:`~DataFrame.apply`
+3. Elementwise_ function application: :meth:`~DataFrame.applymap`
```
this needs backticks to pick up the references

TomAugspurger replied:
backticks on the "Elementwise_"?  It works w/o the backticks since it's a single word. Are do you mean the method references?

jreback replied:
oh, ok, then didn't know that.

jreback commented on doc/source/basics.rst:
```diff
@@ -624,6 +624,77 @@ We can also pass infinite values to define the bins:
+2. `Row or Column-wise Function Application`_: :meth:`~DataFrame.apply`
+3. Elementwise_ function application: :meth:`~DataFrame.applymap`
+
+.. _pipe:
```
``_basics.pipe``

jreback commented on doc/source/basics.rst:
```diff
@@ -624,6 +624,77 @@ We can also pass infinite values to define the bins:
+
+.. versionadded:: 0.16.2
+
+DataFrames and Series can of course just be passed into functions.
```
maybe backtick these (and above too), to make them code like

jreback commented on doc/source/basics.rst:
```diff
@@ -690,6 +762,8 @@ functionality.
    functionality for grouping by some criterion, applying, and combining the
    results into a Series, DataFrame, etc.
 
+.. _Elementwise:
```
``basics.elementwise``

jreback commented on doc/source/basics.rst:
```diff
@@ -624,6 +624,77 @@ We can also pass infinite values to define the bins:
+.. _magrittr: https://github.com/smbache/magrittr
+.. _R: http://www.r-project.org
+
+
```
``basics.apply`` (though I don't think used anywhere ATM)

jreback commented on doc/source/whatsnew/v0.16.2.txt:
```diff
@@ -22,6 +22,56 @@ Check the :ref:`API Changes <whatsnew_0162.api>` before updating.
 New features
 ~~~~~~~~~~~~
 
+We've introduced a new method :meth:`DataFrame.pipe`. As suggested by the name, ``pipe``
```
add a link and ``:ref:`` from the ``Highlites`` section

jreback commented on pandas/core/generic.py:
```diff
@@ -2044,6 +2044,58 @@ def sample(self, n=None, frac=None, replace=False, weights=None, random_state=No
+        pandas.DataFrame.apply
+        pandas.DataFrame.applymap
+        pandas.Series.map
+    """)
```
I think you need a blank line at the end here

TomAugspurger replied:
I think we're good

```python
pd.DataFarme.pipe?
<...>
pandas.DataFrame.apply
pandas.DataFrame.applymap
pandas.Series.map
File:      ~/Envs/dev/lib/python3.4/site-packages/pandas/pandas/core/generic.py
Type:      function
```

jorisvandenbossche replied:
It's indeed not needed, but just a style matter. In the numpydoc format often a blank line is included before the ending triple quote (although we don't always follow that strictly)

jreback commented on pandas/tests/test_generic.py:
```diff
@@ -1649,6 +1649,34 @@ def test_describe_raises(self):
             tm.makePanel().describe()
 
+    def test_pipe(self):
+        df = DataFrame({'A': [1, 2, 3]})
```
Can you add a test with a Series/Panel (of some sort just to validate)

shoyer commented:
FYI monkey patching docs were added by @y-p in this PR: #2530

I am also +1 on removing them... but perhaps someone thinks they are a good idea?

TomAugspurger commented:
Updated...

added panel tests, section in whatsnew/Highligts. Rebuilding the docs now

jreback commented:
ok on removing monkey-patching docs. if you really really want to keep them, then could move to http://pandas.pydata.org/pandas-docs/stable/internals.html#subclassing-pandas-data-structures (FYI need to remove that link as well)

TomAugspurger commented:
Ok, fixed that newline in the docstring and I just link people to the `pipe` section from `internals.rst` instead of the monkeypathcing.

jreback commented on doc/source/whatsnew/v0.16.2.txt:
```diff
@@ -10,6 +10,7 @@ We recommend that all users upgrade to this version.
 Highlights include:
 
 - Documentation on how to use ``numba`` with *pandas*, see :ref:`here <enhancingperf.numba>`
+- A new ``pipe`` method, :ref:`here <basics.pipe>`
```
minor point, but I usually have these refer to the whatsnew itself, e.g. below (with the actual doc link from the whatsnew to the docs)

TomAugspurger replied:
I think the numba one is linking to the docs (could be wrong). Want me to change that to the section in whatsnew?

TomAugspurger replied:
nvm, didn't see it doesn't have a section

jreback replied:
the numba one doesn't have another what's entry that's why its that way. Whereas you have a section in the whatsnew (which is good). So the link will skip to there. Then you already have the link back to the docs (at the end).

jreback commented:
@TomAugspurger lgtm. merge when ready (see that minor point above though)

shoyer commented on doc/source/whatsnew/v0.16.2.txt:
```diff
@@ -22,6 +23,56 @@ Check the :ref:`API Changes <whatsnew_0162.api>` before updating.
+
+In the example above, the functions ``f``, ``g``, and ``h`` each expected the DataFrame as the first positional argument.
+When the funciton you wish to apply takes its data anywhere other than the first argument, pass a tuple
+of ``(funciton, keyword)`` indicating where the DataFrame should flow. For example:
```
spelling: function

shoyer commented on pandas/core/generic.py:
```diff
@@ -2044,6 +2044,59 @@ def sample(self, n=None, frac=None, replace=False, weights=None, random_state=No
+        You can write
+
+        >>> (df.pipe(h)
+               .pipe(g, arg1=a)
```
I think you need to continue lines with `...` to get numpydoc/sphinx to render examples correctly, e.g., 
```
        >>> (df.pipe(h) 
        ...    .pipe(g, arg1=a)
        ...    .pipe(f, arg2=b, arg3=c)
        ... )
```

shoyer commented:
looks pretty much good to go for me, too, after a few quick fixes

TomAugspurger commented:
Ok, I added a section header for `pipe`, link to that from above. Fixed the (two) typos on function, and added the `...` continuations in the docstring.

TomAugspurger commented:
We waiting on Travis? It's felt slow today :/

jreback commented on doc/source/api.rst:
```diff
@@ -1,1597 +0,0 @@
-.. currentmodule:: pandas
```
? I think you checked in after you buildt the docs....!

TomAugspurger replied:
Bah, one sec.

jreback commented:
travis IS slow today (on master anyhow). not real sure why.

TomAugspurger commented:
OK, fixed the accidental deletion of api.rst

jreback commented on pandas/core/generic.py:
```diff
@@ -2045,6 +2045,59 @@ def sample(self, n=None, frac=None, replace=False, weights=None, random_state=No
+        ...    .pipe(g, arg1=a)
+        ...    .pipe(f, arg2=b, arg3=c)
+        ... )
+
```
maybe show an example of using the callable & data_keyword in the Notes? (can do later)

TomAugspurger replied:
Added.

TomAugspurger commented:
@shoyer @jreback merge at your leisure. Or I'll be back online later tonight to push the button :)

shoyer commented on pandas/core/generic.py:
```diff
@@ -2045,6 +2045,68 @@ def sample(self, n=None, frac=None, replace=False, weights=None, random_state=No
+        ... )
+
+        If you have a function that takes the data as (say) the second
+        argumnet, pass a tuple indicating which keyword expects the
```
spelling: argument

TomAugspurger replied:
I'm normally not the bad at spelling.

shoyer replied:
"the bad"? :)

TomAugspurger replied:
Ugh. Long day.

shoyer commented on pandas/core/generic.py:
```diff
@@ -2045,6 +2045,68 @@ def sample(self, n=None, frac=None, replace=False, weights=None, random_state=No
+
+        If you have a function that takes the data as (say) the second
+        argumnet, pass a tuple indicating which keyword expects the
+        data. Say ``f`` takes the data as ``arg2``.
```
Instead of "Say", I would use "For example, if"

This line should also end with a colon `:`, not a period.

jreback commented:
http://blog.ibis-project.org/design-composability/

from @wesm
